File: corpus/beam/megafile/edge_cases.ex
defmodule EdgeCases do
  @moduledoc "Megafile exercising Elixir constructs for BEAM decompilation tests."

  @default_opts %{verbose: false, retries: 3, timeout: 5_000}

  defstruct [:id, :name, :email, tags: [], created_at: nil, meta: %{}]

  @type result(t) :: {:ok, t} | {:error, term()}
  @type id :: non_neg_integer()

  defmodule Address do
    @moduledoc false
    @enforce_keys [:street, :city]
    @derive {Inspect, only: [:city, :country]}
    defstruct [:street, :city, :zip, :country]
  end

  defmodule Greeter do
    @moduledoc false

    @callback greet(binary()) :: binary()
    @callback farewell(binary()) :: binary()

    defmacro __using__(_opts) do
      quote do
        @behaviour Greeter

        def hello(name), do: greet(name)
      end
    end
  end

  defprotocol Renderable do
    @fallback_to_any true
    def render(value)
  end

  defimpl Renderable, for: BitString do
    def render(s), do: "string:" <> s
  end

  defimpl Renderable, for: Integer do
    def render(n), do: "int:#{n}"
  end

  defimpl Renderable, for: List do
    def render(l), do: "list:" <> Enum.join(Enum.map(l, &Renderable.render/1), ",")
  end

  defimpl Renderable, for: Any do
    def render(_), do: "unknown"
  end

  def main do
    IO.puts("EdgeCases main")
    _ = pattern_match_basic({:ok, 42})
    _ = pipe_chain([1, 2, 3, 4, 5])
    _ = with_demo(%{a: 1, b: 2})
    :ok
  end

  def pattern_match_basic({:ok, value}), do: {:matched_ok, value}
  def pattern_match_basic({:error, reason}), do: {:matched_err, reason}
  def pattern_match_basic({:partial, [head | _tail]}), do: {:head, head}
  def pattern_match_basic(atom) when is_atom(atom), do: {:atom, atom}
  def pattern_match_basic(_), do: :no_match

  def pattern_match_map(%{type: :user, id: id, name: name}), do: {:user, id, name}
  def pattern_match_map(%{type: :admin} = m), do: {:admin, m}
  def pattern_match_map(%{} = empty) when map_size(empty) == 0, do: :empty
  def pattern_match_map(_), do: :other

  def pattern_match_struct(%__MODULE__{id: id, name: name}), do: {id, name}
  def pattern_match_struct(%Address{city: city}), do: {:addr, city}
  def pattern_match_struct(_), do: :no_struct

  def pipe_chain(list) do
    list
    |> Enum.map(&(&1 * 2))
    |> Enum.filter(&(&1 > 3))
    |> Enum.reduce(0, fn x, acc -> x + acc end)
  end

  def pipe_with_named(list) do
    list
    |> Enum.map(fn x -> x * x end)
    |> Enum.sort(&>=/2)
    |> Enum.take(3)
    |> Enum.sum()
  end

  def with_demo(map) do
    with {:ok, a} <- Map.fetch(map, :a),
         {:ok, b} <- Map.fetch(map, :b),
         true <- a + b > 0 do
      {:ok, a + b}
    else
      :error -> {:error, :missing_key}
      false -> {:error, :nonpositive}
      _ -> {:error, :unknown}
    end
  end

  def comprehension_simple(list) do
    for x <- list, x > 0, do: x * x
  end

  def comprehension_multi(xs, ys) do
    for x <- xs, y <- ys, x != y, do: {x, y}
  end

  def comprehension_into_map(pairs) do
    for {k, v} <- pairs, is_atom(k), is_integer(v), into: %{}, do: {k, v}
  end

  def comprehension_into_binary(list) do
    for n <- list, n in 0..255, into: <<>>, do: <<n::8>>
  end

  def stream_demo(list) do
    list
    |> Stream.map(&(&1 + 1))
    |> Stream.filter(&(rem(&1, 2) == 0))
    |> Stream.take(5)
    |> Enum.to_list()
  end

  def case_demo(x) do
    case x do
      {:ok, n} when n > 0 -> {:positive, n}
      {:ok, _} -> :zero_or_negative
      {:error, reason} -> {:err, reason}
      _ -> :default
    end
  end

  def cond_demo(x) do
    cond do
      x < 0 -> :negative
      x == 0 -> :zero
      x < 10 -> :small
      true -> :big
    end
  end

  def if_demo(x) do
    if x > 100 do
      :big
    else
      if x > 0 do
        :small
      else
        :zero_or_negative
      end
    end
  end

  def unless_demo(x) do
    unless x, do: :falsy, else: :truthy
  end

  def named_args_demo(name, opts \\ []) do
    verbose = Keyword.get(opts, :verbose, false)
    retries = Keyword.get(opts, :retries, 3)
    {name, verbose, retries}
  end

  def default_args(a, b \\ 10, c \\ 20) do
    a + b + c
  end

  def first_class_callable(fun, list) when is_function(fun, 1) do
    Enum.map(list, fun)
  end

  def capture_demo do
    add = &(&1 + &2)
    inc = &(&1 + 1)
    upcase = &String.upcase/1
    {add.(1, 2), inc.(5), upcase.("hi")}
  end

  def anonymous_multi_clause do
    fn
      {:ok, v} -> v
      {:error, _} -> nil
      :timeout -> :retry
    end
  end

  def map_update(m, k) do
    Map.update(m, k, 1, fn v -> v + 1 end)
  end

  def map_get_in(data) do
    get_in(data, [:user, :addr, :city])
  end

  def map_put_in(data) do
    put_in(data, [:user, :addr, :zip], "12345")
  end

  def map_update_in(data) do
    update_in(data, [:user, :count], &((&1 || 0) + 1))
  end

  def keyword_demo do
    kw = [a: 1, b: 2, c: 3]
    sum = kw |> Keyword.values() |> Enum.sum()
    keys = Keyword.keys(kw)
    {sum, keys}
  end

  def list_ops(list) do
    {Enum.sum(list), Enum.max(list), Enum.min(list), Enum.uniq(list)}
  end

  def tuple_destructure({a, b, c}), do: {c, b, a}
  def tuple_destructure({a, b, c, d}), do: {d, c, b, a}
  def tuple_destructure(t) when is_tuple(t), do: Tuple.to_list(t)

  def binary_pattern(<<0x89, "PNG", _rest::binary>>), do: :png
  def binary_pattern(<<0xFF, 0xD8, _rest::binary>>), do: :jpeg
  def binary_pattern(<<"GIF87a", _rest::binary>>), do: :gif87
  def binary_pattern(<<"GIF89a", _rest::binary>>), do: :gif89
  def binary_pattern(_), do: :unknown

  def binary_construct(tag, payload) when is_atom(tag) and is_binary(payload) do
    tag_bin = Atom.to_string(tag)
    tag_len = byte_size(tag_bin)
    len = byte_size(payload)
    <<1::8, tag_len::8, tag_bin::binary, len::32-big, payload::binary>>
  end

  def string_interp(name, age), do: "Hello, #{name}, you are #{age}"

  def sigil_demo do
    {~w(one two three), ~r/\d+/, ~s("quoted"), ~c"charlist", ~D[2026-05-25]}
  end

  def heredoc_demo do
    """
    Line one
    Line two
    Line three
    """
  end

  defmacro unless_macro(cond, do: do_block, else: else_block) do
    quote do
      if !unquote(cond), do: unquote(do_block), else: unquote(else_block)
    end
  end

  defmacro debug(expr) do
    quote do
      result = unquote(expr)
      IO.inspect(result, label: unquote(Macro.to_string(expr)))
      result
    end
  end

  def use_genserver do
    {:ok, pid} = GenServer.start_link(__MODULE__.MyServer, %{})
    GenServer.call(pid, :get)
  end

  defmodule MyServer do
    @moduledoc false
    use GenServer

    @impl true
    def init(state), do: {:ok, state}

    @impl true
    def handle_call(:get, _from, state), do: {:reply, state, state}

    @impl true
    def handle_call({:put, k, v}, _from, state), do: {:reply, :ok, Map.put(state, k, v)}

    @impl true
    def handle_cast({:inc, k}, state), do: {:noreply, Map.update(state, k, 1, &(&1 + 1))}

    @impl true
    def handle_info(_msg, state), do: {:noreply, state}
  end

  defmodule MySupervisor do
    @moduledoc false
    use Supervisor

    def start_link(arg) do
      Supervisor.start_link(__MODULE__, arg, name: __MODULE__)
    end

    @impl true
    def init(_arg) do
      children = [
        {EdgeCases.MyServer, %{}}
      ]

      Supervisor.init(children, strategy: :one_for_one)
    end
  end

  def agent_demo do
    {:ok, pid} = Agent.start_link(fn -> 0 end)
    Agent.update(pid, &(&1 + 1))
    val = Agent.get(pid, & &1)
    Agent.stop(pid)
    val
  end

  def task_demo do
    task = Task.async(fn -> 1 + 2 end)
    Task.await(task)
  end

  def task_async_stream(list) do
    list
    |> Task.async_stream(fn n -> n * 2 end, max_concurrency: 4)
    |> Enum.map(fn {:ok, v} -> v end)
  end

  def process_demo do
    pid = spawn(fn -> receive do msg -> {:got, msg} end end)
    send(pid, :hello)
    pid
  end

  def spawn_link_demo do
    spawn_link(fn -> :timer.sleep(10) end)
  end

  def spawn_monitor_demo do
    {pid, ref} = spawn_monitor(fn -> :ok end)

    receive do
      {:DOWN, ^ref, :process, ^pid, reason} -> {:done, reason}
    after
      1_000 -> :timeout
    end
  end

  def try_demo(x) do
    try do
      div(10, x)
    rescue
      ArithmeticError -> :divzero
      e in RuntimeError -> {:runtime, e.message}
    catch
      :throw, v -> {:throw, v}
      :exit, r -> {:exit, r}
    else
      n when n > 0 -> {:positive, n}
      n -> {:nonpositive, n}
    after
      :always_runs
    end
  end

  def throw_demo do
    try do
      Enum.each(1..100, fn n ->
        if n == 7, do: throw({:found, n})
      end)

      :not_found
    catch
      {:found, n} -> {:got, n}
    end
  end

  def raise_demo, do: raise(ArgumentError, message: "boom")

  def reraise_demo do
    try do
      raise "inner"
    rescue
      e -> reraise(e, __STACKTRACE__)
    end
  end

  def fibers_like_demo do
    fun = fn _ -> :crypto.strong_rand_bytes(8) end
    Enum.map(1..5, fun)
  end

  def weakmap_like do
    map = %{a: 1, b: 2}
    Enum.into(map, %{})
  end

  def first_class_demo do
    fns = [&String.upcase/1, &String.downcase/1, &String.reverse/1]
    Enum.map(fns, fn f -> f.("hello") end)
  end

  def conditional_compile do
    if function_exported?(:erlang, :system_info, 1) do
      :erlang.system_info(:otp_release)
    else
      :unknown
    end
  end

  def receive_demo do
    receive do
      {:hello, from} -> send(from, :hi)
      :stop -> :stopped
    after
      100 -> :timeout
    end
  end

  def selective_receive do
    receive do
      {:priority, p} -> {:p, p}
    after
      0 ->
        receive do
          other -> {:other, other}
        after
          100 -> :none
        end
    end
  end

  def map_match_deep(%{user: %{addr: %{city: city, zip: zip}}}) do
    {city, zip}
  end

  def map_match_deep(_), do: :no_match

  def function_capture_in_pipe(list) do
    list
    |> Enum.map(&Integer.to_string/1)
    |> Enum.map(&String.pad_leading(&1, 3, "0"))
  end

  def generator_with_filter(list) do
    for x <- list, y <- list, x < y, into: [], do: {x, y}
  end

  def into_struct_demo do
    %__MODULE__{id: 1, name: "alice", tags: [:admin]}
  end

  def update_struct(%__MODULE__{} = ec) do
    %{ec | name: String.upcase(ec.name), meta: Map.put(ec.meta, :updated, true)}
  end

  def access_demo(map) do
    map[:nested][:key]
  end

  def behaviour_user do
    defmodule UserGreeter do
      @moduledoc false
      use Greeter

      def greet(name), do: "Hi, " <> name
      def farewell(name), do: "Bye, " <> name
    end

    UserGreeter.hello("world")
  end
end

File: corpus/beam/megafile/edge_cases.erl
-module(edge_cases).

-behaviour(gen_server).

-export([
    main/0,
    start_link/0,
    start_link/1,
    stop/0,
    init/1,
    handle_call/3,
    handle_cast/2,
    handle_info/2,
    terminate/2,
    code_change/3,
    format_status/2,
    bit_syntax_decode/1,
    bit_syntax_encode/1,
    binary_comprehension/1,
    list_comprehension/1,
    list_comprehension_filtered/1,
    nested_comprehensions/2,
    pattern_match_args/1,
    pattern_match_args/2,
    guarded_dispatch/1,
    record_ops/0,
    map_ops/0,
    map_update/1,
    nested_map_match/1,
    tuple_pivot/1,
    string_ops/1,
    char_list_ops/1,
    fun_capture/1,
    higher_order/2,
    fold_examples/1,
    foldl_with_anon/1,
    spawn_demo/0,
    spawn_link_demo/0,
    spawn_monitor_demo/0,
    receive_demo/0,
    receive_after/1,
    send_pattern/2,
    selective_receive/0,
    ets_demo/0,
    dets_demo/0,
    try_demo/1,
    try_catch_of_after/1,
    exception_chain/1,
    error_with_stacktrace/0,
    process_dict_demo/0,
    hot_code_reload/0,
    apply_demo/3,
    catch_old_school/1,
    proc_lib_demo/0,
    registered_name_demo/0,
    monitor_node_demo/0,
    funlist_idx/2,
    multi_clause_recur/1,
    bifs_demo/0,
    big_int_demo/0,
    float_arith/0,
    boolean_short_circuit/2,
    if_demo/1,
    case_demo/1,
    cond_like/2,
    string_concat_three/3,
    deeply_nested/1,
    deep_pattern_destructure/1
]).

-record(state, {
    count = 0 :: non_neg_integer(),
    name :: atom() | binary() | undefined,
    pids = [] :: [pid()],
    meta = #{} :: #{atom() => term()},
    started_at :: erlang:timestamp() | undefined
}).

-record(person, {
    name :: binary(),
    age :: non_neg_integer(),
    email :: binary() | undefined,
    tags = [] :: [atom()]
}).

-record(point3, {x = 0.0 :: float(), y = 0.0 :: float(), z = 0.0 :: float()}).

-type kv() :: {atom(), term()}.
-type result(T) :: {ok, T} | {error, term()}.

-spec main() -> ok.
main() ->
    io:format("edge_cases main~n", []),
    _ = record_ops(),
    _ = map_ops(),
    _ = bit_syntax_decode(<<1, 2, 3, 4>>),
    _ = list_comprehension([1, 2, 3, 4, 5]),
    ok.

start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [default], []).

start_link(InitArg) ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [InitArg], []).

stop() ->
    gen_server:call(?MODULE, stop).

-spec init(list()) -> {ok, #state{}} | ignore.
init([_InitArg]) ->
    process_flag(trap_exit, true),
    State = #state{
        count = 0,
        name = my_module,
        pids = [],
        meta = #{started => true, version => 1},
        started_at = os:timestamp()
    },
    {ok, State};
init(_) ->
    ignore.

handle_call(stop, _From, State) ->
    {stop, normal, ok, State};
handle_call({add, N}, _From, State = #state{count = C}) when is_integer(N), N > 0 ->
    {reply, ok, State#state{count = C + N}};
handle_call(get_count, _From, State = #state{count = C}) ->
    {reply, {ok, C}, State};
handle_call({set_meta, K, V}, _From, State = #state{meta = M}) ->
    {reply, ok, State#state{meta = M#{K => V}}};
handle_call(get_state, _From, State) ->
    {reply, State, State}.

handle_cast({inc, N}, State = #state{count = C}) when is_integer(N) ->
    {noreply, State#state{count = C + N}};
handle_cast(reset, State) ->
    {noreply, State#state{count = 0, pids = []}};
handle_cast(_, State) ->
    {noreply, State}.

handle_info({'EXIT', From, Reason}, State) ->
    io:format("~p exited: ~p~n", [From, Reason]),
    {noreply, State};
handle_info({'DOWN', _Ref, process, _Pid, _Reason}, State) ->
    {noreply, State};
handle_info(timeout, State) ->
    {noreply, State};
handle_info(_, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

format_status(_Opt, [_PDict, State]) ->
    [{state, State}].

bit_syntax_decode(<<A:8, B:16/big, C:32/little, Rest/binary>>) ->
    {A, B, C, byte_size(Rest)};
bit_syntax_decode(<<Single:8>>) ->
    {single, Single};
bit_syntax_decode(<<>>) ->
    empty.

bit_syntax_encode({Tag, Payload}) when is_atom(Tag), is_binary(Payload) ->
    Len = byte_size(Payload),
    TagBin = atom_to_binary(Tag, utf8),
    TagLen = byte_size(TagBin),
    <<1:8, TagLen:8, TagBin/binary, Len:32/big-unsigned, Payload/binary>>.

binary_comprehension(Bin) when is_binary(Bin) ->
    << <<X:8>> || <<X:8>> <= Bin, X rem 2 =:= 0 >>.

list_comprehension(List) ->
    [X * 2 || X <- List, X > 0, X rem 2 =:= 1].

list_comprehension_filtered(Pairs) ->
    [{K, V} || {K, V} <- Pairs, is_atom(K), is_integer(V), V > 0].

nested_comprehensions(Xs, Ys) ->
    [{X, Y} || X <- Xs, Y <- Ys, X =/= Y].

pattern_match_args({ok, V}) -> {result, V};
pattern_match_args({error, R}) -> {fail, R};
pattern_match_args(_) -> unknown.

pattern_match_args([H | _], all_first) -> H;
pattern_match_args([_, X | _], second) -> X;
pattern_match_args(Atom, _) when is_atom(Atom) -> Atom.

guarded_dispatch(X) when is_integer(X), X > 1000 -> big;
guarded_dispatch(X) when is_integer(X), X > 0 -> small;
guarded_dispatch(X) when is_float(X) -> floaty;
guarded_dispatch(X) when is_binary(X), byte_size(X) > 0 -> bin;
guarded_dispatch(X) when is_list(X), length(X) > 0 -> nonempty_list;
guarded_dispatch(_) -> other.

record_ops() ->
    P0 = #person{name = <<"alice">>, age = 30},
    P1 = P0#person{age = 31, tags = [admin, beta]},
    Name = P1#person.name,
    Age = P1#person.age,
    #person{tags = Tags} = P1,
    {Name, Age, Tags}.

map_ops() ->
    M0 = #{one => 1, two => 2},
    M1 = M0#{three => 3, four => 4},
    M2 = maps:put(five, 5, M1),
    M3 = maps:remove(one, M2),
    Keys = lists:sort(maps:keys(M3)),
    Vals = lists:sort(maps:values(M3)),
    {Keys, Vals}.

map_update(M = #{count := C}) ->
    M#{count => C + 1};
map_update(M) ->
    M#{count => 1}.

nested_map_match(#{outer := #{inner := V}}) -> {ok, V};
nested_map_match(_) -> error.

tuple_pivot({A, B, C}) -> {C, B, A};
tuple_pivot({A, B, C, D}) -> {D, C, B, A};
tuple_pivot(T) -> T.

string_ops(S) ->
    Lower = string:lowercase(S),
    Upper = string:uppercase(S),
    Trimmed = string:trim(S),
    Split = string:split(S, " ", all),
    {Lower, Upper, Trimmed, Split}.

char_list_ops(L) when is_list(L) ->
    Sum = lists:sum([N || N <- L, is_integer(N)]),
    Sorted = lists:sort(L),
    Rev = lists:reverse(L),
    {Sum, Sorted, Rev}.

fun_capture(N) ->
    Plus = fun(X) -> X + N end,
    Mul = fun(X) -> X * N end,
    {Plus, Mul, Plus(10), Mul(10)}.

higher_order(F, List) when is_function(F, 1) ->
    [F(X) || X <- List].

fold_examples(List) ->
    Sum = lists:foldl(fun(X, Acc) -> Acc + X end, 0, List),
    Prod = lists:foldl(fun(X, Acc) -> Acc * X end, 1, List),
    Max = lists:foldl(fun(X, Acc) when X > Acc -> X; (_, Acc) -> Acc end, hd(List), tl(List)),
    {Sum, Prod, Max}.

foldl_with_anon(L) ->
    F = fun
        ({tag, V}, {Tags, Vals}) -> {[V | Tags], Vals};
        (V, {Tags, Vals}) when is_integer(V) -> {Tags, [V | Vals]};
        (_, Acc) -> Acc
    end,
    lists:foldl(F, {[], []}, L).

spawn_demo() ->
    Pid = spawn(fun() -> receive M -> io:format("got: ~p~n", [M]) end end),
    Pid ! hello,
    Pid.

spawn_link_demo() ->
    Pid = spawn_link(fun() -> receive stop -> ok end end),
    Pid.

spawn_monitor_demo() ->
    {Pid, Ref} = spawn_monitor(fun() -> ok end),
    receive
        {'DOWN', Ref, process, Pid, Reason} -> {done, Reason}
    after 1000 ->
        timeout
    end.

receive_demo() ->
    receive
        {greet, From} -> From ! {hi, self()}, ok;
        stop -> stopped
    end.

receive_after(Ms) ->
    receive
        Msg -> {got, Msg}
    after Ms ->
        timeout
    end.

send_pattern(Pid, Msg) ->
    Pid ! Msg,
    {sent, Pid, Msg}.

selective_receive() ->
    receive
        {priority, P} -> {p, P}
    after 0 ->
        receive
            Other -> {other, Other}
        after 100 ->
            none
        end
    end.

ets_demo() ->
    T = ets:new(demo_tab, [set, public]),
    true = ets:insert(T, {alpha, 1}),
    true = ets:insert(T, {beta, 2}),
    [{alpha, 1}] = ets:lookup(T, alpha),
    Size = ets:info(T, size),
    true = ets:delete(T),
    Size.

dets_demo() ->
    File = "demo.dets",
    {ok, _} = dets:open_file(demo_dets, [{file, File}, {type, set}]),
    ok = dets:insert(demo_dets, {key, 42}),
    [{key, 42}] = dets:lookup(demo_dets, key),
    ok = dets:close(demo_dets),
    file:delete(File),
    ok.

try_demo(X) ->
    try
        Y = 10 div X,
        {ok, Y}
    catch
        error:badarith -> {error, divzero};
        Class:Reason -> {error, {Class, Reason}}
    end.

try_catch_of_after(X) ->
    try compute(X) of
        N when is_integer(N), N > 0 -> {positive, N};
        N -> {nonpositive, N}
    catch
        throw:Tag -> {throw, Tag};
        error:Reason -> {err, Reason};
        exit:Reason -> {exit_, Reason}
    after
        ok
    end.

exception_chain(0) ->
    throw(zero);
exception_chain(N) when N < 0 ->
    error({negative, N});
exception_chain(N) ->
    exit({too_big, N}).

error_with_stacktrace() ->
    try
        error(forced)
    catch
        error:Reason:Stack -> {Reason, length(Stack) > 0}
    end.

process_dict_demo() ->
    put(my_key, 42),
    V = get(my_key),
    erase(my_key),
    V.

hot_code_reload() ->
    code:purge(?MODULE),
    code:load_file(?MODULE),
    ?MODULE:main().

apply_demo(M, F, A) ->
    apply(M, F, A).

catch_old_school(Op) ->
    case catch Op() of
        {'EXIT', R} -> {error, R};
        V -> {ok, V}
    end.

proc_lib_demo() ->
    Pid = proc_lib:spawn_link(fun() -> ok end),
    Pid.

registered_name_demo() ->
    Pid = spawn(fun() -> receive _ -> ok end end),
    true = register(named_demo, Pid),
    named_demo ! stop,
    unregister(named_demo),
    ok.

monitor_node_demo() ->
    erlang:monitor_node(node(), true),
    erlang:monitor_node(node(), false),
    ok.

funlist_idx([], _) -> none;
funlist_idx([F | _], 0) -> {ok, F};
funlist_idx([_ | T], N) when N > 0 -> funlist_idx(T, N - 1).

multi_clause_recur(0) -> base;
multi_clause_recur(1) -> one;
multi_clause_recur(N) when N rem 2 =:= 0 -> {even, multi_clause_recur(N - 2)};
multi_clause_recur(N) -> {odd, multi_clause_recur(N - 2)}.

bifs_demo() ->
    A = erlang:atom_to_list(hello),
    B = erlang:integer_to_binary(12345),
    C = erlang:list_to_tuple([1, 2, 3]),
    D = erlang:pid_to_list(self()),
    {A, B, C, D}.

big_int_demo() ->
    A = 999999999999999999999,
    B = A * A,
    {A, B, B > A}.

float_arith() ->
    A = 1.5 + 2.25,
    B = 3.14 * 2,
    C = math:pi() / 2.0,
    {A, B, C}.

boolean_short_circuit(X, Y) ->
    A = X andalso Y,
    B = X orelse Y,
    C = X and Y,
    D = X or Y,
    {A, B, C, D}.

if_demo(X) ->
    if
        X > 100 -> big;
        X > 10 -> medium;
        X > 0 -> small;
        true -> other
    end.

case_demo(X) ->
    case X of
        {ok, _} -> matched_ok;
        {error, _} -> matched_err;
        _ when is_atom(X) -> atom_branch;
        _ -> default
    end.

cond_like(A, B) ->
    case {A, B} of
        {true, _} -> first;
        {_, true} -> second;
        _ -> none
    end.

string_concat_three(A, B, C) ->
    <<A/binary, B/binary, C/binary>>.

deeply_nested(L) ->
    case L of
        [[[V | _] | _] | _] -> {deep, V};
        [[V | _] | _] -> {two, V};
        [V | _] -> {one, V};
        [] -> empty
    end.

deep_pattern_destructure({user, #{name := N, addr := #{city := City, zip := Zip}}}) ->
    {N, City, Zip};
deep_pattern_destructure(_) ->
    no_match.

compute(0) -> throw(zero);
compute(N) when N > 0 -> N * 2;
compute(N) -> -N.
